### PR TITLE
feat(secrets): Add multiline detection to entropy keyword combinator

### PR DESCRIPTION
**Pull request by marynaKK**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

[//]: # "
    # PR Title
    Be aware that we use the title to create changelog automatically and therefore only allow specific prefixes
    - break:    to indicate a breaking change, this supersedes any of the types
    - feat:     to indicate new features or checks
    - fix:      to indicate a bugfix or handling of edge cases of existing checks
    - docs:     to indicate an update to our documentation
    - chore:    to indicate adjustments to workflow files or dependency updates
    - platform: to indicate a change needed for the platform
    Additionally a scope is needs to be added to the prefix, which indicates the targeted framework, in doubt choose 'general'.
    ex.
    feat(terraform): add CKV_AWS_123 to ensure that VPC Endpoint Service is configured for Manual Acceptance
"

## Description

* changed the logic of parsing yml files (part of changes of IAC secret detection): secret and keyword appearing in the context of a line should be also parsed as a pair, and the secret should be detected.

for IAC files only keyword combination is supported. this addition supports also multiline combination.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my feature, policy, or fix is effective and works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/secrets/plugins/entropy_keyword_combinator.py [hunk 1]
```diff
@@ -15,6 +32,61 @@
 ENTROPY_KEYWORD_COMBINATOR_LIMIT = 3
 ENTROPY_KEYWORD_LIMIT = 4.5
 
+INDENTATION_PATTERN = re.compile(r'(^[\r\t\f\v ]*(?:-?[\r\t\f\v ]+)?)')
+COMMENT_PREFIX = re.compile(r'^[\r\t\f\v ]*#')
```
Comment (nimrodkor) [inline checkov/secrets/plugins/entropy_keyword_combinator.py:36]: `\s` character is for all whitespace, can replace these `\r\t\f\v `
Also, // is also a comment. So:
```suggestion
INDENTATION_PATTERN = re.compile(r'(^\s*(?:-?\s+)?)')
COMMENT_PREFIX = re.compile(r'^[\s]*(#|\/\/)')
```

File: checkov/secrets/plugins/entropy_keyword_combinator.py [hunk 2]
```diff
@@ -34,26 +109,125 @@ def analyze_line(
...
+            line = context.lines[j]
+            if self.lines_in_same_object(raw_context=raw_context, idx=j) \
+                    and not self.line_is_comment(line):
+                possible_keywords.add(raw_context.lines[j])
+                if self.is_object_start(raw_context=raw_context, idx=j):
+                    return possible_keywords
+        return possible_keywords
+
+    @staticmethod
+    def format_reducing_noice_secret(string):
```
Comment (nimrodkor) [inline checkov/secrets/plugins/entropy_keyword_combinator.py:187]:
```suggestion
    def format_reducing_noise_secret(string):
```

File: checkov/secrets/plugins/entropy_keyword_combinator.py [hunk 3]
```diff
@@ -34,26 +109,125 @@ def analyze_line(
...
+
+        # classic key-value pair
+        keyword_on_key = self.keyword_scanner.analyze_line(filename, line, line_number, **kwargs)
+        if keyword_on_key:
+            return self.detect_secret(self.high_entropy_scanners_iac, filename, line, line_number, **kwargs)
+
+        # not so classic key-value pair, from multiline, that is only in an array format.
+        # The scan is one-way backwards, so no duplicates expected.
+
+        elif context is not None and raw_context is not None:
```
Comment (nimrodkor) [inline checkov/secrets/plugins/entropy_keyword_combinator.py:154]: @marynaKK Everything below this line is for yaml ONLY. Need to make sure that non yaml files are not affected by this change at all - please make sure that non-yaml files are run the same way they did before. It is not clear from the flow that this is what would happen.

File: tests/secrets/resources/cfn/test-multiline-secrets.yml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+spec:
```
Comment (nimrodkor) [inline tests/secrets/resources/cfn/test-multiline-secrets.yml:1]: @marynaKK perhaps move to a separate dir to verify you don't affect other tests?

**Review by nimrodkor — COMMENTED** [review]
@marynaKK Very nice job!

File: checkov/secrets/plugins/entropy_keyword_combinator.py (hunk 3 above)
Comment (marynaKK) [reply checkov/secrets/plugins/entropy_keyword_combinator.py:154]: Checked if it’s yml but forgot to use it - great to have a review!!!

Comment (marynaKK) [reply checkov/secrets/plugins/entropy_keyword_combinator.py]: added some logic - so now only yaml should be parsed.

**Review by nimrodkor — APPROVED** [review]
We will move all yaml related methods to OOP to allow easier implementations for other extensions in another PR. GJ!

File: checkov/secrets/plugins/entropy_keyword_combinator.py (hunk 2 above)
Comment (gruebel) [reply checkov/secrets/plugins/entropy_keyword_combinator.py]: damn, I liked the `noice` method 😄

File: checkov/secrets/plugins/entropy_keyword_combinator.py [hunk 5]
```diff
@@ -34,26 +108,151 @@ def analyze_line(
...
+        return False
+
+    @staticmethod
+    def line_is_comment(line: str) -> bool:
+        if re.match(COMMENT_PREFIX, line):
+            return True
+        return False
+
+    @staticmethod
+    def extract_from_string(pattern: Optional[dict[Pattern[str], int]], string: str) -> str:
```
Comment (gruebel) [inline checkov/secrets/plugins/entropy_keyword_combinator.py:223]:
```suggestion
    def extract_from_string(pattern: dict[Pattern[str], int] | None, string: str) -> str:
```

File: checkov/secrets/plugins/entropy_keyword_combinator.py [hunk 6]
```diff
@@ -34,26 +108,151 @@ def analyze_line(
...
+                forward_range = range(context.target_index - 1, -1, -1)
+                backwards_range = range(context.target_index + 1, len(context.lines))
+                possible_keywords |= self.get_lines_from_same_object(forward_range, context, raw_context)
+                possible_keywords |= self.get_lines_from_same_object(backwards_range, context, raw_context)
+
+                for other_value in possible_keywords:
+                    if self.extract_from_string(pattern=value_pattern, string=other_value):
+                        secrets |= entropy_on_value
+                    if secrets:
+                        break
```
Comment (gruebel) [inline checkov/secrets/plugins/entropy_keyword_combinator.py:170]:
```suggestion
                        secrets |= entropy_on_value
                        break
```
should be the same or am I missing something?

**Review by gruebel — COMMENTED** [review]
nice, work, just have some minor comments

File: checkov/secrets/plugins/entropy_keyword_combinator.py (hunk 6 above)
Comment (marynaKK) [reply checkov/secrets/plugins/entropy_keyword_combinator.py]: changed it all! :) thanks for the detailed review!

**Review by gruebel — APPROVED** [review]
🐶

**Pull request merged**